src/main/java/com/github/cameltooling/lsp/internal/parser/CamelKModelineInsertionParser.java

```diff
@@ -0,0 +1,72 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.github.cameltooling.lsp.internal.parser;
+
+import com.github.cameltooling.lsp.internal.completion.modeline.CamelKModelineFileType;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.TextDocumentItem;
+
+import java.util.Arrays;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Parses and checks text document to see if camel-k modeline insertion completions are available.
+ *
+ * @author joshiraez
+ */
+public class CamelKModelineInsertionParser {
+
+    private final TextDocumentItem document;
+
+    public CamelKModelineInsertionParser(TextDocumentItem document) {
+        this.document = document;
+    }
+
+    public boolean canPutCamelKModeline(Position position) {
+        int currentLine = position.getLine();
+
+        return isCamelKFile()
+                && lineIsEmpty(currentLine)
+                && noModelineInsertedAlready()
+                && previousLinesAreCommentsOrEmpty(currentLine);
+    }
+
+    private boolean isCamelKFile() {
+        return CamelKModelineFileType.getFileTypeCorrespondingToUri(document.getUri()).isPresent();
+    }
+
+    private boolean lineIsEmpty(int line) {
+        return new ParserFileHelperUtil().getLine(document, line).isBlank();
+    }
+
+    private boolean noModelineInsertedAlready() {
+        String modeline = CamelKModelineFileType.getFileTypeCorrespondingToUri(document.getUri()).orElseThrow()
+                .modeline;
+
+        return !Arrays.stream(document.getText().split("\n")).anyMatch(line -> line.startsWith(modeline));
```

Comment:
After reading the cod more carefully I think it would be simpler to check for the optional one time (as it is anyway checked) and avoid using orElseThrow(). It is currently checked in isCamelKFile method so it is dead code. Making it clear that it is checked seems easier to follow the flow when reading the code.
something like:

```

    public boolean canPutCamelKModeline(Position position) {
        int currentLine = position.getLine();
        Optional<CamelKModelineFileType> isCamelKFile = CamelKModelineFileType.getFileTypeCorrespondingToUri(document.getUri());
        if (isCamelKFile.isPresent()) {
        	return lineIsEmpty(currentLine)
                    && noModelineInsertedAlready(isCamelKFile.get())
                    && previousLinesAreCommentsOrEmpty(currentLine, isCamelKFile.get());
        } else {
        	return false;
        }
    }

    private boolean lineIsEmpty(int line) {
        return new ParserFileHelperUtil().getLine(document, line).isBlank();
    }

    private boolean noModelineInsertedAlready(CamelKModelineFileType camelKModelineFileType) {
        String modeline =camelKModelineFileType.modeline;

        return Arrays.stream(document.getText().split("\n")).noneMatch(line -> line.startsWith(modeline));
    }

    private boolean previousLinesAreCommentsOrEmpty(int line, CamelKModelineFileType camelKModelineFileType) {
        String textBeforeLine = IntStream.range(0, line).boxed()
                .map(currLine -> new ParserFileHelperUtil().getLine(document,currLine))
                .collect(Collectors.joining("\n"));

        return camelKModelineFileType.checkTextIsCommentsDelegate.apply(textBeforeLine);
    }
    
  ```

What do you think?

Reply:
I think that makes a lot of sense.

Reply:
I took a more functional approach using Optional.map(conditionals).orElse because I think better used Optional capabilities and leads to more readable code, but other than that I copied your idea :)

Reply:
I applied the same idea on the processor to get rid of orElseThrow there as well